osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java

```diff
@@ -76,24 +82,35 @@ public void run() {
         LOGGER.info("Attempting key recovery for device {}", this.deviceIdentification);
 
         try {
-            this.initDevice();
+            this.findDevice();
         } catch (final Exception e) {
-            LOGGER.error("Unexpected exception: {}", e);
+            LOGGER.error("Could not find device", e);
+            //TODO why try to find device if you don't do anything with the result?!?
+            //TODO return?
         }
-        if (!this.device.hasNewSecurityKey()) {
+
+        if (!this.secretManagementService.hasNewSecretOfType(this.deviceIdentification, E_METER_AUTHENTICATION)) {
+            LOGGER.warn("Could not recover keys: device has no new authorisation key registered in secret-mgmt module");
             return;
         }
 
-        if (this.canConnect()) {
-            this.promoteInvalidKey();
+        if (this.canConnectUsingNewKeys()) {
+            List<SecurityKeyType> keyTypesToActivate=Arrays.asList(E_METER_ENCRYPTION,E_METER_AUTHENTICATION);
+            try {
+                this.secretManagementService.activateNewKeys(this.deviceIdentification, keyTypesToActivate);
+            } catch (Exception e) {
+                throw new RecoverKeyException(e.getMessage(), e);
```

Comment:
Why not _throw new RecoverKeyException(e);_

Reply:
Agreed.